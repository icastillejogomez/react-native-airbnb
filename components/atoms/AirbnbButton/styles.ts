import { Platform, StyleSheet } from 'react-native'

export const styles = StyleSheet.create({
  container: {
    height: Platform.select({ ios: 54, android: 52 }),
    borderRadius: 8,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
  contained: {},
  outlined: {
    borderWidth: 1,
  },
})
