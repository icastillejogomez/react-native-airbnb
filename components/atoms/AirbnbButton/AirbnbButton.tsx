import { AirbnbText } from '@/components/native'
import { forwardRef, useCallback } from 'react'
import { View, Pressable, PressableProps, StyleProp, ViewStyle, GestureResponderEvent } from 'react-native'
import Animated, { interpolate, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'

import { styles } from './styles'
import { usePalette } from '@/theme'

// Create pressable animated component
const AnimatedPressable = Animated.createAnimatedComponent(Pressable)

export type AirbnbButtonVariant = 'contained' | 'outlined'

export type AirbnbButtonProps = Omit<PressableProps, 'style'> & {
  variant?: AirbnbButtonVariant
  title: string
  style?: ViewStyle
}

const AirbnbButton = forwardRef<View, AirbnbButtonProps>((props, ref) => {
  const { title, variant = 'contained', style: parentStyle, onPressIn, onPressOut, ...rest } = props

  const palette = usePalette()
  const isScaled = useSharedValue(0)

  const style: StyleProp<ViewStyle> = [
    styles.container,
    variant === 'contained' ? [{ ...styles.contained }, { backgroundColor: palette.background['primary-core'] }] : styles.outlined,
    parentStyle || {},
  ]

  const handlePressIn = useCallback(
    (e: GestureResponderEvent) => {
      isScaled.value = withTiming(1)
      onPressIn?.(e)
    },
    [onPressIn, isScaled],
  )

  const handlePressOut = useCallback(
    (e: GestureResponderEvent) => {
      isScaled.value = withTiming(0)
      onPressOut?.(e)
    },
    [onPressOut, isScaled],
  )

  const animatedStyle = useAnimatedStyle(() => {
    return {
      transform: [{ scale: interpolate(isScaled.value, [0, 1], [1, 0.95], 'clamp') }],
    }
  })

  return (
    <AnimatedPressable style={[style, animatedStyle]} {...rest} ref={ref} onPressIn={handlePressIn} onPressOut={handlePressOut}>
      <AirbnbText
        color={variant === 'contained' ? 'primary-inverse' : 'primary'}
        weight={variant === 'contained' ? '600' : '500'}
        variant="body1"
      >
        {title}
      </AirbnbText>
    </AnimatedPressable>
  )
})
AirbnbButton.displayName = 'AirbnbButton'

export { AirbnbButton }
