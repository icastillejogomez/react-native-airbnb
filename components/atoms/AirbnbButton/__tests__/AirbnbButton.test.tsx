import { render, userEvent } from '@testing-library/react-native'
import { AirbnbButton } from '../AirbnbButton'

jest.useFakeTimers()

describe('AirbnbButton', () => {
  it('renders without crashing', () => {
    const tree = render(<AirbnbButton title="Airbnb" />)
    expect(tree).toMatchSnapshot()
  })

  it('render the title on it', () => {
    const tree = render(<AirbnbButton title="Airbnb" />)
    expect(tree.getByText('Airbnb')).toBeTruthy()
  })

  it('calls onPress when pressed', async () => {
    const onPress = jest.fn()
    const tree = render(<AirbnbButton testID="button" title="Airbnb" onPress={onPress} />)

    userEvent.setup({ advanceTimers: jest.advanceTimersByTime })
    const $element = tree.getByTestId('button')
    await userEvent.press($element)
    expect(onPress).toHaveBeenCalledTimes(1)
  })
})
