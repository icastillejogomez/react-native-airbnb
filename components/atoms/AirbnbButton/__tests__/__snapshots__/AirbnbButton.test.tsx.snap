// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirbnbButton renders without crashing 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  jestAnimatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FF385C",
      "borderRadius": 8,
      "flexDirection": "row",
      "height": 54,
      "justifyContent": "center",
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
        },
        {
          "color": "#FFFFFF",
        },
        {
          "fontFamily": "Montserrat-Medium",
          "fontSize": 16,
        },
        {
          "fontFamily": "Montserrat-SemiBold",
        },
        {},
        {},
        {
          "textAlign": "auto",
        },
        {},
        {},
      ]
    }
  >
    Airbnb
  </Text>
</View>
`;
